### ISSUE TITLE
Dealing with multiple child refs

### ISSUE THREAD

**Issue by joepio** [onderonur/react-intersection-observer-hook#18]
Hi, thanks for creating this!

So I want to track a list of items (chat messages in an infinite scrolling list) in a custom scrollable div. If I understand correctly, this means I should use `rootRef` for the parent. However, since I have multiple children, I need a `ref` for every chilld, but only one for the parent. Does this library support this? I tried setting the `root` in the options (passing the `parentRef`), but for some reason that does not work.

**Comment by ancorgs** [comment]
Exactly same problem/situation here.

**Comment by onderonur** [comment]
Hello there 👋 

Firstly, hugely sorry for responding too late 🤦 I didn't have the chance to resolve this issue unfortunately. Currently, I'm migrating my packages to a new structure to make it easier to maintain them. So, I finally had the chance to resolve this. Again, sorry for that 🙏 

Tracking multiple child components with the same `rootRef` is not very easy with the current design of `useIntersectionObserver` hook.

I have a possible suggestion for this, but it is not bulletproof. If the root element changes, it may have some issues.

So, let's say we have a `<Message>` component like this:
```tsx
import { useEffect, useRef } from "react";
import { useIntersectionObserver } from "react-intersection-observer-hook";

export type MessageData = {
  id: string;
  content: string;
};

export type MessageProps = {
  data: MessageData;
  parentRef: React.RefObject<HTMLUListElement | null>;
  onVisibilityChange: (isVisible: boolean) => void;
};

export function Message({ data, parentRef, onVisibilityChange }: MessageProps) {
  const [ref, { entry, rootRef }] = useIntersectionObserver();

  useEffect(() => {
    rootRef(parentRef.current);
  }, [parentRef, rootRef]);

  const savedCallbackRef = useRef<MessageProps["onVisibilityChange"]>();

  useEffect(() => {
    savedCallbackRef.current = onVisibilityChange;
  }, [onVisibilityChange]);

  useEffect(() => {
    const callback = savedCallbackRef.current;

    if (callback) {
      callback(!!entry?.isIntersecting);
    }
  }, [entry]);

  return (
    <div ref={ref}>
      {data.content} {entry?.rootBounds?.height}
    </div>
  );
}
```

It accepts `parentRef` prop which can be used for passing the `ref` of root element to be observed by `IntersectionObserver`.

And we can use this component like below.
```tsx
import { useRef, useState } from "react";
import { Message, MessageData } from "./message";

const messages = [
  {
    id: "1",
    content: "Do consectetur labore culpa labore commodo pariatur.",
  },
  {
    id: "2",
    content:
      "In aute sint fugiat aliquip sit elit quis adipisicing officia exercitation nulla et dolore Lorem.",
  },
  {
    id: "3",
    content: "Pariatur dolore culpa magna ad ut esse.",
  },
  {
    id: "4",
    content:
      "Dolore culpa incididunt magna et pariatur fugiat consectetur dolore sint in.",
  },
  {
    id: "5",
    content: "Laborum pariatur occaecat velit velit in.",
  },
];

export function App() {
  const [visibleMessages, setVisibleMessages] = useState<MessageData[]>([]);
  const parentRef = useRef<HTMLUListElement>(null);

  return (
    <main>
      <p>Visible Message Count: {visibleMessages.length}</p>
      <ul ref={parentRef} style={{ maxHeight: 300, overflow: "auto" }}>
        {messages.map((message) => {
          return (
            <li key={message.id} style={{ minHeight: 120 }}>
              <Message
                data={message}
                parentRef={parentRef}
                onVisibilityChange={(isVisible) => {
                  setVisibleMessages((current) => {
                    if (isVisible) {
                      return [...current, message];
                    }
                    return current.filter((m) => m.id !== message.id);
                  });
                }}
              />
            </li>
          );
        })}
      </ul>
    </main>
  );
}
```

So, we set `parentRef` by passing it to `<ul>`, which is the parent element of multiple `<Message>` components.

And, `parentRef` is passed to each `<Message>` component as a prop and in the `useEffect` of each `<Message>`, we start observing according to `<ul>` as the `root`.

This can work for some cases. But since it is not possible to re-run `useEffect` according to `ref` changes (even if you pass `ref.current` as a prop, it might not be possible for some cases) there can be edge cases which behaves in unexpected ways.

Another way (which is a little more complex to show in a comment) might be storing `rootRef`s of each `<Message>` component in a state of their parent component (also not forgetting to unregister them) and combining these `rootRef` callbacks into a single callback and passing it to `<ul>` as a `ref`.

But again, I think it can have some edge cases too and is more complex than it should be.

In an ideal case, `useIntersectionObserver` might provide an API to observe multiple children components under the same `root`. But currently it is not possible.

If I change the API of this hook and make this possible, I will definitely update this issue. But for now, I will close this.

**TLDR:**  This package does not support this behavior. It is possible, but can have a lot of edge cases.
